### PR TITLE
Add TotalScore Snapshot

### PR DESCRIPTION
**Pull request by SvetoslavUzunov**
Closes: https://github.com/SoftUni-Internal/exam-systems-issues/issues/821

Added two new properties to Participant entity: 
int TotalScoreSnapshot
Datetime? TotalScoreSnapshotModifiedOn

These properties will be updated/calculated everytime user submits a submission.

Two new Background jobs added.

Background job that will update participant total score once per day. Background job that will remove Participant scores if there are more than one for specific problem , for practice or compete. It will remain only the one with highest points.

### CODE REVIEW

File: Data/Infrastructure/OJS.Data.Infrastructure/Extensions/DbContextOptionsBuilderExtensions.cs [hunk 1]
```diff
@@ -18,7 +18,9 @@ public static class DbContextOptionsBuilderExtensions
             // For migrations
             // var connectionString = "Server=.;Database=OpenJudgeSystem;Integrated Security=True;";
             // var connectionString = "Server=.;Database=OpenJudgeSystem;User Id=sa;Password=1234;";
-            var connectionString = EnvironmentUtils.GetApplicationConnectionString(applicationName);
+            //var connectionString = EnvironmentUtils.GetApplicationConnectionString(applicationName);
+
+            var connectionString = "Server=.;Database=OpenJudgeSystem;Integrated Security=True;User Id=sa;Password=1234;";
```
Comment (MTsatsarov) [inline Data/Infrastructure/OJS.Data.Infrastructure/Extensions/DbContextOptionsBuilderExtensions.cs:24]: Revert changes.

File: Data/OJS.Data/OJS.Data.csproj [hunk 2]
```diff
@@ -11,6 +11,15 @@
...
       <PackageReference Include="Microsoft.AspNetCore.DataProtection.EntityFrameworkCore" Version="6.0.16" />
+      <PackageReference Include="Microsoft.EntityFrameworkCore.Design" Version="6.0.16">
+        <PrivateAssets>all</PrivateAssets>
+        <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
+      </PackageReference>
+      <PackageReference Include="Microsoft.EntityFrameworkCore.SqlServer" Version="6.0.16" />
+      <PackageReference Include="Microsoft.EntityFrameworkCore.Tools" Version="6.0.16">
+        <PrivateAssets>all</PrivateAssets>
+        <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
+      </PackageReference>
```
Comment (MTsatsarov) [inline Data/OJS.Data/OJS.Data.csproj:22]: Why are these packages added ?

File: Servers/Administration/OJS.Servers.Administration/OJS.Servers.Administration.csproj [hunk 3]
```diff
@@ -8,6 +8,7 @@
     <ItemGroup>
       <ProjectReference Include="..\..\..\Services\Administration\OJS.Services.Administration.Business\OJS.Services.Administration.Business.csproj" />
       <ProjectReference Include="..\..\..\Services\Administration\OJS.Services.Administration.Data\OJS.Services.Administration.Data.csproj" />
+      <ProjectReference Include="..\..\..\Services\UI\OJS.Services.Ui.Business\OJS.Services.Ui.Business.csproj" />
```
Comment (MTsatsarov) [inline Servers/Administration/OJS.Servers.Administration/OJS.Servers.Administration.csproj:11]: Same here

File: Services/UI/OJS.Services.Ui.Business/Implementations/ParticipantScoresBusinessService.cs [hunk 4]
```diff
@@ -182,12 +184,13 @@ private async Task NormalizeSubmissionPoints()
         private async Task NormalizeParticipantScorePoints()
             => await (await this.participantScoresData
                     .GetAllHavingPointsExceedingLimit()
-                    .Select(ps => new { ParticipantScore = ps, ProblemMaxPoints = ps.Problem.MaximumPoints, })
+                    .Select(ps => new { ParticipantScore = ps, ProblemMaxPoints = ps.Problem.MaximumPoints, Participant=ps.Participant })
```
Comment (MTsatsarov) [inline Services/UI/OJS.Services.Ui.Business/Implementations/ParticipantScoresBusinessService.cs:187]: Fix spacing

File: Services/UI/OJS.Services.Ui.Data/Implementations/ParticipantScoresDataService.cs [hunk 5]
```diff
@@ -69,6 +70,7 @@ public async Task ResetBySubmission(Submission submission)
                 {
                     p.IsOfficial,
                     p.User.UserName,
+                    Participant=p,
```
Comment (MTsatsarov) [inline Services/UI/OJS.Services.Ui.Data/Implementations/ParticipantScoresDataService.cs:73]: Fix spacing

File: Services/Common/OJS.Services.Common.Data/Implementations/DataService.cs [hunk 6]
```diff
@@ -152,5 +151,22 @@ public virtual IQueryable<TEntity> GetByIdQuery(object id)
 
             return query;
         }
+
+        public async Task ExecuteSqlCommandWithTimeout(string query, int timeoutInSeconds)
+        {
```
Comment (MTsatsarov) [inline Services/Common/OJS.Services.Common.Data/Implementations/DataService.cs:156]: Should be added also to the interface. Add description what means to be executed with timeoutInSeconds = 0

**Review by MTsatsarov — CHANGES_REQUESTED** [review]

File: Servers/UI/OJS.Servers.Ui/Controllers/Api/ParticipantScoresController.cs [hunk 7]
```diff
@@ -13,8 +13,8 @@ public class ParticipantScoresController : BaseApiController
 {
     private IParticipantScoresBusinessService participantScoresBusinessService;
 
-    public ParticipantScoresController(IParticipantScoresBusinessService participantScoresBusinessService)
-        => this.participantScoresBusinessService = participantScoresBusinessService;
+    public ParticipantScoresController(IParticipantScoresBusinessService participantScoresBusinessService) =>
```
Comment (PetarNeshkov) [inline Servers/UI/OJS.Servers.Ui/Controllers/Api/ParticipantScoresController.cs:16]: Revert '=>' as before

File: Services/UI/OJS.Services.Ui.Business/Implementations/ParticipantsBusinessService.cs [hunk 8]
```diff
@@ -41,6 +44,28 @@ public Task<int> GetPracticeParticipantsCount(int contestId)
     public Task<int> GetCompeteParticipantsCount(int contestId)
         => this.participantsData.GetAllByContestAndIsOfficial(contestId, true).CountAsync();
 
+    public async Task UpdateTotalScoreSnapshotOfParticipants() =>
```
Comment (PetarNeshkov) [inline Services/UI/OJS.Services.Ui.Business/Implementations/ParticipantsBusinessService.cs:47]: Place '=>' on the new line. This is a convention.

File: Services/UI/OJS.Services.Ui.Business/Implementations/ParticipantsBusinessService.cs [hunk 9]
```diff
@@ -41,6 +44,28 @@ public Task<int> GetPracticeParticipantsCount(int contestId)
...
+    public async Task RemoveParticipantMultipleScores()
+    {
+        var participantScores =
+            this.scoresDataService.GetAll()
+                .GroupBy(ps => new { ps.IsOfficial, ps.ProblemId, ps.ParticipantId })
+                .Where(ps => ps.Count() > 1)
+                .ToList();
+
+        var participantScoresToRemove = new List<ParticipantScore>();
+        foreach (var participantScoreGroup in participantScores)
```
Comment (PetarNeshkov) [inline Services/UI/OJS.Services.Ui.Business/Implementations/ParticipantsBusinessService.cs:59]: Try to use .ForEach() whenever you can. This is the better approach.

File: Services/UI/OJS.Services.Ui.Business/Implementations/ParticipantsBusinessService.cs [hunk 10]
```diff
@@ -41,6 +44,28 @@ public Task<int> GetPracticeParticipantsCount(int contestId)
...
+        {
+            participantScoresToRemove
+                .AddRange(participantScoreGroup.OrderByDescending(ps => ps.Points).Skip(1)
```
Comment (PetarNeshkov) [inline Services/UI/OJS.Services.Ui.Business/Implementations/ParticipantsBusinessService.cs:62]: create a separate variable for the input:
`participantScoreGroup.OrderByDescending(ps => ps.Points)
                                      .Skip(1)`

Also, fix the the indentation:
`participantScoresToRemove.AddRange()
                                              .ToList()`;

**Review by PetarNeshkov — CHANGES_REQUESTED** [review]
Requesting changes

File: Servers/Administration/OJS.Servers.Administration/Controllers/BackgroundJobsController.cs [hunk 11]
```diff
@@ -51,4 +51,26 @@ public IActionResult AddRecurringDeletingOldSubmissionsForProcessingJob()
...
+    [HttpGet]
+    public ActionResult? RegisterJobForRemovingMultipleParticipantScoresForProblem()
+    {
+        this.hangfireBackgroundJobsService.AddOrUpdateRecurringJob<IParticipantsBusinessService>(
+            "RemoveParticipantMultipleScores",
+            p => p.RemoveParticipantMultipleScores(),
+            Cron.Daily(3));
+
+        return null;
+    }
```
Comment (gogo4ds) [inline Servers/Administration/OJS.Servers.Administration/Controllers/BackgroundJobsController.cs:75]: We have another way of registering jobs - see `BackgroundJobsHostedService`

**Review by gogo4ds — CHANGES_REQUESTED** [review]

File: Services/Administration/OJS.Services.Administration.Business/Implementations/RecurringBackgroundJobsBusinessService.cs [hunk 12]
```diff
@@ -6,9 +6,15 @@
     public class RecurringBackgroundJobsBusinessService : IRecurringBackgroundJobsBusinessService
     {
         private readonly ISubmissionsForProcessingBusinessService submissionsForProcessing;
+        private readonly OJS.Services.Ui.Business.IParticipantsBusinessService participantsBusinessService;
 
-        public RecurringBackgroundJobsBusinessService(ISubmissionsForProcessingBusinessService submissionsForProcessing)
-            => this.submissionsForProcessing = submissionsForProcessing;
+        public RecurringBackgroundJobsBusinessService(
+            ISubmissionsForProcessingBusinessService submissionsForProcessing,
+            OJS.Services.Ui.Business.IParticipantsBusinessService participantsBusinessService)
```
Comment (gogo4ds) [inline Services/Administration/OJS.Services.Administration.Business/Implementations/RecurringBackgroundJobsBusinessService.cs:13]: Administration should not reference Ui. Move logic into Administration's `IParticipantsBusinessService`.

File: Services/Administration/OJS.Services.Administration.Models/ParticipantScores/Models/CategoryContestsParticipationSummary.cs [hunk 13]
```diff
@@ -0,0 +1,10 @@
+﻿namespace OJS.Services.Administration.Models.ParticipantScores.Models;
+
+using System.Collections.Generic;
+
+public class CategoryContestsParticipationSummary
```
Comment (gogo4ds) [inline Services/Administration/OJS.Services.Administration.Models/ParticipantScores/Models/CategoryContestsParticipationSummary.cs:5]: What is this used for?

**Review by gogo4ds — CHANGES_REQUESTED** [review]

**Review by MTsatsarov — APPROVED** [review]

File: Services/Administration/OJS.Services.Administration.Models/ParticipantScores/Models/CategoryContestsParticipationSummary.cs (hunk 13 above)
Comment (SvetoslavUzunov) [reply Services/Administration/OJS.Services.Administration.Models/ParticipantScores/Models/CategoryContestsParticipationSummary.cs]: it's fixed

**Pull request merged**